components/formats-bsd/test/spec/schema/Schema_Transform_Test.java

```diff
@@ -0,0 +1,830 @@
+        for (int j = 0; j < map.getLength(); j++) {
+            attribute = map.item(j);
+            schema = attribute.getNodeValue();
+            transforms = node.getElementsByTagName(TRANSFORM);
+            list = new ArrayList<String>();
+            for (int i = 0; i < transforms.getLength(); i++) {
+                NamedNodeMap m = transforms.item(i).getAttributes();
+                for (int k = 0; k < m.getLength(); k++) {
+                    attribute = m.item(k);
+                    if (FILE.equals(attribute.getNodeName()))
+                        list.add(attribute.getNodeValue());
+                }
+            }
+        }
+        if (!schema.trim().isEmpty() && !list.isEmpty()) {
+            sheets.put(schema, list);
+        }
+    }
+
+    /**
+     * Extracts the value of the current schema.
+     *
+     * @param schema The current value.
+     * @throws Exception Thrown when an error occurred while parsing the file.
+     */
+    private Map<String, List<String>> extractCurrentSchema(String schema,
+            Document document)
+        throws Exception
+    {
+        NodeList list = document.getElementsByTagName(SOURCE);
+        Element n;
+        Node attribute;
+        NamedNodeMap map;
+        NodeList t;
+        Map<String, List<String>> transforms =
+                new HashMap<String, List<String>>();
+        Map<String, List<String>> umap;
+        for (int i = 0; i < list.getLength(); ++i) {
+            n = (Element) list.item(i);
+            map = n.getAttributes();
+            for (int j = 0; j < map.getLength(); j++) {
+                attribute = map.item(j);
+                if (SCHEMA.equals(attribute.getNodeName())) {
+                    if (schema.equals(attribute.getNodeValue())) {
+                        t = n.getElementsByTagName(TARGET);
+                        for (int k = 0; k < t.getLength(); k++) {
+                            parseTarget((Element) t.item(k), transforms);
+                        }
+                    } else {
+                        NodeList nl = n.getElementsByTagName("upgrades");
+                        umap = new HashMap<String, List<String>>();
+                        String src = attribute.getNodeValue();
+                        for (int k = 0; k < nl.getLength(); k++) {
+                            Element node = (Element) nl.item(k);
+                            NodeList tt = node.getElementsByTagName(TARGET);
+                            for (int l = 0; l < tt.getLength(); l++) {
+                                parseTarget((Element) tt.item(l), umap);
+                            }
+                            //parse the map
+                            Iterator<Entry<String, List<String>>> kk = 
+                                    umap.entrySet().iterator();
+                            Entry<String, List<String>> e;
+                            while (kk.hasNext()) {
+                                e = kk.next();
```

Comment:
The upgrades map could probably be built directly from umap without the need for the iterator in this case. Initially this code was merely replicating the parsing used in the ExporterTest as closely as possible.